### PR TITLE
36 take into account permissions when editing a user

### PR DESCRIPTION
**Pull request by Maximelego**
Added permission checking method.
Added permission architecture.

The architecture is quite complete.
We basically define CRUD operations and Services.
We define a set of permissions which are the combination of CRUD operations on a specific service.
Roles are assigned to users and contains some numbers of permissions (combinations of operations CRUD on services).

Took the opportunity of this branch to move around some code that didn't belong in the UserInDB model

Added Typing to ORM models.

To check if a user has the permission on a route, simply call the check_permission method, with the current user (an example of how to get it is in the PATCH user Route)

For now the permissions are not representative and are defined in dummy JSON files. Plan to change this in the future.
By default, every user is an admin. This is to make create an update routes not a mess.

Please examine this PR with care, it is a critical feature that still needs more testing.

### CODE REVIEW

**Comment by desaleo** [comment]
i just hope i didn't break anything while merging

File: app/models/tortoise/user.py [hunk 1]
```diff
@@ -2,94 +2,33 @@
...
-    lastname = fields.CharField(max_length=128)
-    mail = fields.CharField(unique=True, required=True, max_length=128)
-    hash = fields.TextField(required=True)
+    id        : Field[int]                   = IntField(primary_key=True)
+    login     : Field[str]                   = CharField(unique=True, required=True, max_length=128)
+    firstname : Field[str]                   = CharField(max_length=128)
+    lastname  : Field[str]                   = CharField(max_length=128)
+    mail      : Field[str]                   = CharField(unique=True, required=True, max_length=128)
+    hash      : Field[str]                   = TextField(required=True)
+    role      : ForeignKeyRelation[RoleInDB] = ForeignKeyField("models.RoleInDB", related_name="role", default="Administrator")
```
Comment (HugoBien) [inline app/models/tortoise/user.py:24]: idk if let admin be the default role is the best choice, i think put the role with the less responsabilites is better

File: app/services/PermissionService.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    """
+    This method checks if the provided user has the permission to perform the provided operation on the provided service.
+    """
+    # We fetch the user's role.
+    await current_user.fetch_related("role")
+    role: RoleInDB = current_user.role
+
+    # We fetch the permissions of the role.
+    # We also filter the permissions to get only the ones that match the service and the operation.
+    permissions: list[PermissionInDB] = await role.permissions.filter(service_name_id=service.value, operation_name_id=operation.value)
```
Comment (HugoBien) [inline app/services/PermissionService.py:23]: I don't understand how u can have more than one permission, if u choose for exemple service: user, operation : delete, u should have only one result ?

**Review by HugoBien — APPROVED** [review]

File: app/models/tortoise/user.py (hunk 1 above)
Comment (Maximelego) [reply app/models/tortoise/user.py:24]: This is only for the first sprint, will be, of course, changed at a later date.
Not even sure if we should set a default role, we should make this field required in creation.

File: app/services/PermissionService.py (hunk 2 above)
Comment (Maximelego) [reply app/services/PermissionService.py:23]: Basically yes, you can have the permission to GET, but not to DELETE for instance.
We will add more operations to define them better (like /GET on me, GET on others, ...)

File: app/models/tortoise/user.py (hunk 1 above)
Comment (HugoBien) [reply app/models/tortoise/user.py:24]: ok !

**Review by desaleo — APPROVED** [review]
LGTM

**Pull request merged**